pkg/linters/forbiddenusage/forbiddenusage.go

```diff
@@ -0,0 +1,103 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package forbiddenUsageprovides is a linter to detect unwanted/deprecated function calls in the code base.
+package forbiddenusage
+
+import (
+	"go/ast"
+	"strings"
+
+	"github.com/golangci/plugin-module-register/register"
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/passes/inspect"
+	"golang.org/x/tools/go/ast/inspector"
+)
+
+func init() {
+	register.Plugin("forbiddenUsage", New)
+}
+
+type forbiddenUsagePlugin struct{}
+
+// New returns a new forbiddenUsage linter plugin
+func New(any) (register.LinterPlugin, error) {
+	return &forbiddenUsagePlugin{}, nil
+}
+
+// BuildAnalyzers returns the analyzers for the plugin
+func (f *forbiddenUsagePlugin) BuildAnalyzers() ([]*analysis.Analyzer, error) {
+	return []*analysis.Analyzer{
+		{
+			Name:     "forbiddenusage",
+			Doc:      "detect unwanted/deprecated function calls",
+			Run:      f.run,
+			Requires: []*analysis.Analyzer{inspect.Analyzer},
+		},
+	}, nil
+}
+
+// GetLoadMode returns the load mode for the plugin
+func (f *forbiddenUsagePlugin) GetLoadMode() string {
+	return register.LoadModeSyntax
+}
+
+// Extend this list to detect more unwanted method call in the code base.
+var forbiddenUsage = map[string]string{
+	"github.com/DataDog/datadog-agent/pkg/config/model.NewConfig": "do not create new config, use 'pkg/config/mock' for test and the config comp otherwise",
+}
+
+func (f *forbiddenUsagePlugin) run(pass *analysis.Pass) (interface{}, error) {
+	inspector := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+
+	// filter needed nodes: imports and function calls
+	nodeFilter := []ast.Node{
+		(*ast.ImportSpec)(nil),
+		(*ast.CallExpr)(nil),
+	}
+
+	importsAliases := map[string]string{}
+
+	inspector.Preorder(nodeFilter, func(node ast.Node) {
+		// Preorder traverse the tree from top to bottom guaranteeing we'll parse imports before function calls.
+
+		// Is the node an import
+		imp, ok := node.(*ast.ImportSpec)
+		if ok {
+			importPath := strings.Trim(imp.Path.Value, "\"")
+
+			// Does the import has an alias
+			if imp.Name != nil {
+				importsAliases[imp.Name.Name] = importPath
+			} else {
+				importName := importPath[strings.LastIndex(importPath, "/")+1:]
+				importsAliases[importName] = importPath
+			}
+			return
+		}
+
+		// Is the node a function call
+		call, ok := node.(*ast.CallExpr)
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏Would this only catch actual function calls ? Eg. would it catch storing the function in a variable (which could be called later) ?
If not it might be better to just check `SelectorExpr` ?

Reply:
We could but if we extend the selector to match for example usage of constant or global vars it would not longer work. I think it's better to keep 1 if case per selector.